pkg/controller/replicaset/replica_set_test.go

```diff
@@ -1312,7 +1319,60 @@ func TestReadyReplicas(t *testing.T) {
 	manager.syncReplicaSet(getKey(rs, t))
 
 	// ReadyReplicas should go from 0 to 2.
-	rs.Status = extensions.ReplicaSetStatus{Replicas: 2, ReadyReplicas: 2, ObservedGeneration: 1}
+	rs.Status = extensions.ReplicaSetStatus{Replicas: 2, ReadyReplicas: 2, AvailableReplicas: 2, ObservedGeneration: 1}
+
+	decRs := runtime.EncodeOrDie(testapi.Extensions.Codec(), rs)
+	fakeHandler.ValidateRequest(t, testapi.Extensions.ResourcePath(replicaSetResourceName(), rs.Namespace, rs.Name)+"/status", "PUT", &decRs)
+	validateSyncReplicaSet(t, &fakePodControl, 0, 0, 0)
+}
+
+var nowFn = func() *unversioned.Time {
+	return nil
+}
+
+func TestAvailableReplicas(t *testing.T) {
+	// This is a happy server just to record the PUT request we expect for status.Replicas
+	fakeHandler := utiltesting.FakeHandler{
+		StatusCode:   200,
+		ResponseBody: "{}",
+	}
+	testServer := httptest.NewServer(&fakeHandler)
+	defer testServer.Close()
+
+	client := clientset.NewForConfigOrDie(&restclient.Config{Host: testServer.URL, ContentConfig: restclient.ContentConfig{GroupVersion: testapi.Default.GroupVersion()}})
+	manager := NewReplicaSetControllerFromClient(client, controller.NoResyncPeriodFunc, BurstReplicas, 0)
+	manager.podStoreSynced = alwaysReady
+
+	// Status.Replica should update to match number of pods in system, 1 new pod should be created.
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newReplicaSet(2, labelMap)
+	rs.Status = extensions.ReplicaSetStatus{Replicas: 2, ReadyReplicas: 0, AvailableReplicas: 0, ObservedGeneration: 1}
+	rs.Generation = 1
+	// minReadySeconds set to 20s
+	rs.Spec.MinReadySeconds = 20
+	manager.rsStore.Store.Add(rs)
+
+	// First pod becomes ready at 00:00
+	moment := unversioned.Date(2016, 11, 13, 0, 0, 0, 0, time.UTC)
+	nowFn = func() *unversioned.Time { return &moment }
+	newPodList(manager.podStore.Indexer, 1, api.PodRunning, labelMap, rs, "pod")
+	// Second pod becomes ready at 00:10
+	anotherMoment := unversioned.Date(2016, 11, 13, 0, 0, 10, 0, time.UTC)
+	nowFn = func() *unversioned.Time { return &anotherMoment }
+	newPodList(manager.podStore.Indexer, 1, api.PodRunning, labelMap, rs, "otherPod")
```

Comment:
`nowFn` is used to set pod's lastTransitionTime. IMHO the name is misleading and may be confused with `now`. Also, the caller after `TestAvailableReplicas` will be affected since `nowFn` is changed. Consider remove it and just passing the time to `newPodList`